src/FakeItEasy.Analyzer/ArgumentConstraintTypeMismatchAnalyzer.cs

```diff
@@ -61,23 +63,62 @@ protected override void AnalyzeArgumentConstraintCore(SyntaxNodeAnalysisContext
             }
 
             var parameter = parameters[index];
-            if (parameter.Type is INamedTypeSymbol parameterType && parameterType.IsNullable())
+            if (parameter.Type is INamedTypeSymbol parameterType)
             {
-                var nonNullableParameterType = parameterType.TypeArguments[0];
-                if (constraintType.Equals(nonNullableParameterType))
+                var nonNullableParameterType = parameterType.IsNullable() ? parameterType.TypeArguments[0] : null;
+
+                if (nonNullableParameterType != null &&
+                    constraintType.IsValueType &&
+                    !constraintType.IsNullable() &&
+                    constraintType.Equals(nonNullableParameterType))
+                {
+                    if (propertyFullName != "FakeItEasy.A`1.That")
+                    {
+                        var descriptor = DiagnosticDefinitions.ArgumentConstraintNullabilityMismatch;
+                        var diagnostic = Diagnostic.Create(
+                            descriptor,
+                            completeConstraint.GetLocation(),
+                            parameter.Name,
+                            methodOrIndexerName,
+                            completeConstraint.ToString());
+                        context.ReportDiagnostic(diagnostic);
+                    }
+                }
+                else if (!ConstraintTypeCanBeAssignedToParameterType(constraintType, parameterType))
                 {
-                    var descriptor = DiagnosticDefinitions.ArgumentConstraintNullabilityMismatch;
+                    var descriptor = DiagnosticDefinitions.ArgumentConstraintTypeMismatch;
                     var diagnostic = Diagnostic.Create(
                         descriptor,
                         completeConstraint.GetLocation(),
                         parameter.Name,
                         methodOrIndexerName,
-                        completeConstraint.ToString());
+                        parameter.Type,
+                        completeConstraint.ToString(),
+                        constraintType);
                     context.ReportDiagnostic(diagnostic);
                 }
             }
         }
 
+        private static bool ConstraintTypeCanBeAssignedToParameterType(INamedTypeSymbol constraintType, INamedTypeSymbol parameterType)
+        {
+            
+            var currentType = constraintType;
+            do
+            {
+                if (currentType.Equals(parameterType) ||
+                    currentType.AllInterfaces.Contains(parameterType))
+                {
+                    return true;
+                }
+
+                currentType = currentType.BaseType;
+            }
+            while (currentType != null);
+
+            return false;
+        }

```

Comment:
Nicely done! Another possible approach would be to use [`ClassifyConversion`](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/CSharpExtensions.cs,1068)

Reply:
I gave it a go. It was a lot of guesswork, and I'm not sure the result made it worthwhile, but I'll let you judge.

Reply:
I like it!